web-api/src/ingestion/routes.rs

```diff
@@ -177,65 +226,87 @@ async fn upsert_documents(
         .into_iter()
         .map(UnvalidatedIngestedDocument::validate)
         .partition_result::<Vec<_>, Vec<_>, _, _>();
-    let snippets = storage::Document::get_excerpted(
+
+    let existing_data = storage::Document::get_excerpted(
         &state.storage,
         documents.iter().map(|document| &document.id),
     )
     .await?
     .into_iter()
-    .map(|document| (document.id, document.snippet))
+    .map(|document| (document.id, (document.snippet, document.is_candidate)))
     .collect::<HashMap<_, _>>();
 
-    let (changed_documents, new_documents) =
-        documents.into_iter().partition::<Vec<_>, _>(|document| {
-            snippets
+    // Hint: Documents which have a changed snippet are also in `new_documents`.
+    let (changed_documents, new_documents) = documents
+        .into_iter()
+        .map(|document| {
+            let (snippet, is_candidate) = existing_data
                 .get(&document.id)
-                .map(|snippet| document.snippet == *snippet)
-                .unwrap_or_default()
-        });
+                .map(|(s, c)| (s, *c))
```

Comment:
:art: unzipping is fine, i would just not abbreviate the names to make it more readable
```suggestion
                .map(|(snippet, is_candidate)| (snippet, *is_candidate))
```